File: js/bin/interop_checks.js
// Corpus-driven interop checks for the JS port: the `framing` subcommand
// confirms the JS frame classifier reaches the same verdicts as the Java,
// Elixir, Rust, and Go implementations over spec/corpus/framing.json, and
// `messages` does the same for the message validator over
// spec/corpus/messages.json. Invoked by interop/check-framing-js.sh and
// interop/check-messages-js.sh.
import fs from 'node:fs';
import { classify, HANDSHAKE_CAP, TRANSPORT_CAP } from '../src/frame.js';
import { validate } from '../src/message.js';
import {
  split,
  MAX_SEGMENT_BYTES,
  MAX_CHUNKS,
  MAX_REASSEMBLY_BUFFER,
  MAX_CONCURRENT_REASSEMBLIES,
  REASSEMBLY_TIMEOUT_MILLIS,
} from '../src/chunk.js';

const [mode, path] = process.argv.slice(2);
if (!mode || !path) {
  process.stderr.write('usage: interop_checks <framing|messages|chunk> <corpus.json>\n');
  process.exit(2);
}
const doc = JSON.parse(fs.readFileSync(path, 'utf8'));

let failures = 0;
const report = (name, ok) => { console.log(`${ok ? 'PASS' : 'FAIL'} ${name}`); if (!ok) failures++; };

if (mode === 'framing') {
  for (const c of doc.cases) {
    const cap = c.kind === 'handshake' ? HANDSHAKE_CAP : TRANSPORT_CAP;
    const { reason } = classify(Buffer.from(c.bytes_b64, 'base64'), cap);
    const ok = c.expect === 'accept' ? reason === undefined : reason === c.reason;
    report(c.name, ok);
  }
  console.log(`framing: ${doc.cases.length} cases checked`);
} else if (mode === 'messages') {
  for (const c of doc.cases) {
    const reason = validate(c.schema, c.frame);
    const ok = c.expect === 'valid' ? reason === null : reason === c.reason;
    report(c.name, ok);
  }
  console.log(`messages: ${doc.cases.length} cases checked`);
} else if (mode === 'chunk') {
  // Two things, and the second is the one nothing else can see. First the pinned
  // §0 constants must match this implementation's -- including the three (chunk
  // count, in-flight count, timeout) that specsrc deliberately does not check,
  // because a substring search for 1024, 256 or 30000 is satisfied by any buffer
  // size already in the tree. Second, the segments this implementation produces
  // must land on exactly the byte boundaries the corpus pins, which is how all
  // seven are shown to cut in the SAME places rather than merely to cut.
  const mine = {
    max_segment_bytes: MAX_SEGMENT_BYTES,
    max_chunks: MAX_CHUNKS,
    max_reassembly_buffer: MAX_REASSEMBLY_BUFFER,
    max_concurrent_reassemblies: MAX_CONCURRENT_REASSEMBLIES,
    reassembly_timeout_millis: REASSEMBLY_TIMEOUT_MILLIS,
  };
  const pinned = doc.constants || {};
  if (Object.keys(pinned).length === 0) {
    process.stderr.write('corpus declares no chunk constants\n');
    process.exit(1);
  }
  for (const [name, want] of Object.entries(pinned)) {
    const ok = mine[name] === want;
    report(`constant ${name}${ok ? '' : `  (have ${mine[name]}, corpus pins ${want})`}`, ok);
  }
  if (!doc.split_cases || doc.split_cases.length === 0) {
    process.stderr.write('corpus has no split cases\n');
    process.exit(1);
  }
  for (const c of doc.split_cases) {
    const payload = { [c.key]: c.unit.repeat(c.times) };
    const msgs = split(doc.mid, 'a', 'b', 16, payload);
    const whole = msgs.length === 1 && 'payload' in msgs[0];
    const lengths = whole ? [] : msgs.map((m) => Buffer.byteLength(m.seg, 'utf8'));
    let ok =
      whole === c.expect_whole &&
      JSON.stringify(lengths) === JSON.stringify(c.segment_byte_lengths);
    let detail = ok
      ? ''
      : `  (whole=${whole} want ${c.expect_whole}; lengths=${JSON.stringify(lengths.slice(0, 8))} want ${JSON.stringify(c.segment_byte_lengths.slice(0, 8))})`;
    // A round-trip as the second oracle: matching lengths would not catch segments
    // that are the right size and the wrong bytes.
    if (ok && !whole) {
      const rebuilt = JSON.parse(msgs.map((m) => m.seg).join(''));
      if (JSON.stringify(rebuilt) !== JSON.stringify(payload)) {
        ok = false;
        detail = '  (segments did not rebuild the payload)';
      }
    }
    report(`${c.name}${detail}`, ok);
  }
  if (failures === 0) console.log('splitting agrees with every pinned constant and cut position');
} else {
  process.stderr.write(`unknown mode: ${mode}\n`);
  process.exit(2);
}

if (failures > 0) process.exit(1);

File: rust/src/bin/interop_checks.rs
//! Corpus-driven interop checks for the Rust port: `framing` and `messages`
//! subcommands confirm the Rust frame classifier and message validator reach
//! the same verdicts as the Java, Go, and Elixir implementations over the shared
//! corpora. Invoked by interop/check-framing-rust.sh and check-messages-rust.sh.

use std::process::exit;

use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use bonemesh::chunk;
use bonemesh::frame::{self, Verdict};
use bonemesh::message;
use serde_json::Value;

fn main() {
    let mode = std::env::args().nth(1).expect("mode: framing|messages|chunk");
    let path = std::env::args().nth(2).expect("corpus path");
    let doc: Value = serde_json::from_str(&std::fs::read_to_string(&path).unwrap()).unwrap();

    let fails = match mode.as_str() {
        "framing" => check_framing(&doc),
        "messages" => check_messages(&doc),
        "chunk" => check_chunk(&doc),
        other => {
            eprintln!("unknown mode: {other}");
            exit(2);
        }
    };
    if fails > 0 {
        exit(1);
    }
}

fn check_framing(doc: &Value) -> usize {
    let mut fails = 0;
    for c in doc["cases"].as_array().unwrap() {
        let name = c["name"].as_str().unwrap();
        let cap = if c["kind"] == "handshake" { frame::HANDSHAKE_CAP } else { frame::TRANSPORT_CAP };
        let raw = B64.decode(c["bytes_b64"].as_str().unwrap()).unwrap();
        let ok = match (c["expect"].as_str().unwrap(), frame::classify(&raw, cap)) {
            ("accept", Verdict::Accept(_)) => true,
            ("reject", Verdict::Reject(r)) => Some(r) == c["reason"].as_str(),
            _ => false,
        };
        report(name, ok, &mut fails);
    }
    println!("framing: {} cases checked", doc["cases"].as_array().unwrap().len());
    fails
}

fn check_messages(doc: &Value) -> usize {
    let mut fails = 0;
    for c in doc["cases"].as_array().unwrap() {
        let name = c["name"].as_str().unwrap();
        let r = message::validate(c["schema"].as_str().unwrap(), &c["frame"]);
        let ok = match c["expect"].as_str().unwrap() {
            "valid" => r.is_none(),
            "invalid" => r == c["reason"].as_str(),
            _ => false,
        };
        report(name, ok, &mut fails);
    }
    println!("messages: {} cases checked", doc["cases"].as_array().unwrap().len());
    fails
}

/// Verifies splitting against the shared corpus (spec/corpus/chunk.json).
///
/// Two things, and the second is the one nothing else can see. First the pinned §0
/// constants must match this implementation's -- including the three (chunk count,
/// in-flight count, timeout) that specsrc deliberately does not check, because a
/// substring search for 1024, 256 or 30000 is satisfied by any buffer size already in
/// the tree. Second, the segments this implementation produces must land on exactly
/// the byte boundaries the corpus pins, which is how all seven are shown to cut in the
/// SAME places rather than merely to cut.
fn check_chunk(doc: &Value) -> usize {
    let mut fails = 0usize;
    let mine: [(&str, i64); 5] = [
        ("max_segment_bytes", chunk::MAX_SEGMENT_BYTES as i64),
        ("max_chunks", chunk::MAX_CHUNKS),
        ("max_reassembly_buffer", chunk::MAX_REASSEMBLY_BUFFER as i64),
        ("max_concurrent_reassemblies", chunk::MAX_CONCURRENT_REASSEMBLIES as i64),
        ("reassembly_timeout_millis", chunk::REASSEMBLY_TIMEOUT_MILLIS),
    ];
    let pinned = doc["constants"].as_object().expect("corpus declares no chunk constants");
    for (name, want) in pinned {
        let want = want.as_i64().unwrap();
        let got = mine.iter().find(|(k, _)| k == name).map(|(_, v)| *v);
        let ok = got == Some(want);
        if ok {
            println!("PASS constant {name}");
        } else {
            println!("FAIL constant {name}  (have {got:?}, corpus pins {want})");
            fails += 1;
        }
    }

    let cases = doc["split_cases"].as_array().expect("corpus has no split cases");
    let mid = doc["mid"].as_str().unwrap();
    for c in cases {
        let name = c["name"].as_str().unwrap();
        let key = c["key"].as_str().unwrap();
        let unit = c["unit"].as_str().unwrap();
        let times = c["times"].as_u64().unwrap() as usize;
        let payload = serde_json::json!({ key: unit.repeat(times) });

        let msgs = match chunk::split(mid, "a", "b", 16, payload.clone()) {
            Ok(m) => m,
            Err(e) => {
                println!("FAIL {name}  (split: {e})");
                fails += 1;
                continue;
            }
        };
        let whole = msgs.len() == 1 && msgs[0].get("payload").is_some();
        let lengths: Vec<usize> = if whole {
            vec![]
        } else {
            msgs.iter().map(|m| m["seg"].as_str().unwrap().len()).collect()
        };
        let want_whole = c["expect_whole"].as_bool().unwrap();
        let want_lengths: Vec<usize> = c["segment_byte_lengths"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_u64().unwrap() as usize)
            .collect();

        let mut ok = whole == want_whole && lengths == want_lengths;
        let mut detail = String::new();
        if !ok {
            detail = format!(
                "  (whole={whole} want {want_whole}; lengths={lengths:?} want {want_lengths:?})"
            );
        }
        // A round-trip as the second oracle: matching lengths would not catch segments
        // that are the right size and the wrong bytes.
        if ok && !whole {
            let joined: String = msgs.iter().map(|m| m["seg"].as_str().unwrap()).collect();
            match serde_json::from_str::<Value>(&joined) {
                Ok(rebuilt) if rebuilt == payload => {}
                _ => {
                    ok = false;
                    detail = "  (segments did not rebuild the payload)".to_string();
                }
            }
        }
        if ok {
            println!("PASS {name}");
        } else {
            println!("FAIL {name}{detail}");
            fails += 1;
        }
    }
    if fails == 0 {
        println!("splitting agrees with every pinned constant and cut position");
    }
    fails
}

fn report(name: &str, ok: bool, fails: &mut usize) {
    if ok {
        println!("PASS {name}");
    } else {
        println!("FAIL {name}");
        *fails += 1;
    }
}
